### ISSUE TITLE
Feature idea: persistent named copy/paste buffers.

### ISSUE THREAD

**Issue by junkyardsparkle** [darktable-org/darktable#3266]
tl;dr: Paste buffers would be used in a similar way to styles, but trading off longevity/portability for lower-level functionality.

**Is your feature request related to a problem? Please describe.**
As noted in #3026 and elsewhere, it was concluded that styles could not safely apply iop ordering. Since I found this capability very useful during the short time it existed, I've been thinking about how this functionality could be more safely reimplemented. For modules such as lut3D, the ability to quickly and easily apply particular module settings *including the position within the pixelpipe* will probably be important to its usefulness (#3108).

**Describe the solution you'd like**
While using my favorite text editor, I had the thought that persistent, named copy/paste buffers could be just as useful within darktable. As currently implemented, "copy" operations simply reference a particular images history stack (or a portion of it), but nothing is actually copied until a following paste operation. This creates limitations in how copy/paste can be used, since things can happen to the reference, leaving you with nothing to paste; therefore current copy/paste really needs to be a closely coupled operation.

If, however, it was also possible to copy the *actual* desired information about the history stack into buffers, which could be saved and applied using a similar interface to styles/presets (including shortcut binding), this would solve the above problem without the need to try to make styles work in a way they were never intended for. These buffers would be part of local config, and *not* for the purpose of exporting and sharing. The logic while applying could be similar to pasting currently, or "smarter" if needed.

**Describe alternatives you've considered**
The most obvious alternative would be to try to safely shoehorn this functionality into styles, but my impression is that this would open another proverbial "can of worms", and seems unlikely to happen soon.

**Additional context**
I'm sure, as with everything else, this idea has issues that I'm not thinking about... please be constructive in pointing them out. ;-)

**Comment by leclercj** [comment]
It would be great indeed to have ability to apply customized iop orders!
But I have to admit I do not understand what's fundamentally different and superior in this proposal compared to the existing styles.

**Comment by junkyardsparkle** [comment]
They would basically be a complement to styles. Styles are a legacy from the fixed-iop-order days, and I suspect they'll need to stay similar to what they are to maintain backwards compatibility (unless they get a versioning system like most other things in darktable). Styles could be described as "an exportable bag of presets", and are expected to be portable, able to be shared and stored independently from a database or other configuration.

The proposed "paste buffers" would give up on that, and would be part of a specific local configuration, like presets. In exchange for this, they would be able to do what styles can not do, essentially working the same way as copy/paste does now, but could be stored as "snapshots" of a copy/paste action instead of being transient. The user interface for creating them would resemble the relevant parts of the user interface for styles and copy/paste.

In other words, they resemble styles in usage, but copy/paste in implementation (and be subject to whatever additional restrictions are needed for "safety").

**Comment by junkyardsparkle** [comment]
Oh, and there are some cases where copy/paste fails now, due to various actions taken in between copying and pasting... so the use case really extends from "slightly more persistent paste source" to something more like the local, non-shared use of styles for quickly applying frequently used chunks of processing.

**Comment by junkyardsparkle** [comment]
It looks like the module ordering issues will be addressed by #3908, so closing this.
